File: examples/statcast_search.py
#!/usr/bin/env python3
"""Statcast Search - batter wOBA/xwOBA CSV download (Python).

Usage: python statcast_search.py [year] [end_date]
Requires: requests, pandas (pip install requests pandas)
"""
import io
import sys
from datetime import date, timedelta

import pandas as pd
import requests

YEAR = sys.argv[1] if len(sys.argv) > 1 else "2025"
END_DATE = sys.argv[2] if len(sys.argv) > 2 else "2025-09-28"

end = date.fromisoformat(END_DATE)
start = end - timedelta(days=6)

params = {
    "hfSea": f"{YEAR}|",
    "hfGT": "R|",
    "player_type": "batter",
    "game_date_gt": start.isoformat(),
    "game_date_lt": end.isoformat(),
    "group_by": "name",
    "sort_col": "xwoba",
    "sort_order": "desc",
    "min_pas": "0",
    "chk_stats_pa": "on",
    "chk_stats_woba": "on",
    "chk_stats_xwoba": "on",
}

print(f"Downloading {YEAR} batter stats ({start} to {end})...")
r = requests.get("https://baseballsavant.mlb.com/statcast_search/csv", params=params, timeout=60)
r.raise_for_status()

df = pd.read_csv(io.StringIO(r.text))
cols = [c for c in ("player_name", "player_id", "pa", "woba", "xwoba") if c in df.columns]
print(df[cols].head(10).to_string(index=False))

File: examples/statcast_search.mjs
// Statcast Search - batter wOBA/xwOBA CSV download (Node.js)
// Usage: node statcast_search.mjs [year] [end_date]
// Requires: csv-parse (npm install csv-parse) - in package.json
import { parse } from "csv-parse/sync";

const YEAR = process.argv[2] || "2025";
const END_DATE = process.argv[3] || "2025-09-28";

const end = new Date(`${END_DATE}T00:00:00Z`);
const start = new Date(end);
start.setUTCDate(start.getUTCDate() - 6);
const startDate = start.toISOString().slice(0, 10);

const url = new URL("https://baseballsavant.mlb.com/statcast_search/csv");
const params = {
  hfSea: `${YEAR}|`,
  hfGT: "R|",
  player_type: "batter",
  game_date_gt: startDate,
  game_date_lt: END_DATE,
  group_by: "name",
  sort_col: "xwoba",
  sort_order: "desc",
  min_pas: "0",
  chk_stats_pa: "on",
  chk_stats_woba: "on",
  chk_stats_xwoba: "on",
};
Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v));

console.log(`Downloading ${YEAR} batter stats (${startDate} to ${END_DATE})...`);
const res = await fetch(url);
if (!res.ok) throw new Error(`Request failed: HTTP ${res.status}`);
const csv = await res.text();

const rows = parse(csv, { columns: true, skip_empty_lines: true, bom: true });
console.log(`Rows: ${rows.length}`);
for (const row of rows.slice(0, 5)) {
  console.log(
    row.player_name,
    `PA=${row.pa}`,
    `wOBA=${row.woba}`,
    `xwOBA=${row.xwoba}`
  );
}